File: cpp/HybridTempoClassifier.cpp
#include "HybridTempoClassifier.hpp"

#include <cstdio>

#if defined(__ANDROID__)
#include <android/log.h>
#endif

namespace {

constexpr const char* kLogTag = "NitroTempoCls";

void logDebug(const char* message) {
#if defined(__ANDROID__)
  __android_log_print(ANDROID_LOG_DEBUG, kLogTag, "%s", message);
#else
  std::fprintf(stderr, "[%s] %s\n", kLogTag, message);
#endif
}

} // namespace

namespace margelo::nitro::tempoclassifier {

bool HybridTempoClassifier::loadModelFromJson(const std::string& /*modelJson*/) {
  logDebug("loadModelFromJson(): iOS fallback - not supported");
  return false;
}

bool HybridTempoClassifier::loadModelFromAsset(const std::string& /*assetName*/) {
  logDebug("loadModelFromAsset(): iOS fallback - not supported");
  return false;
}

void HybridTempoClassifier::setExercise(const std::optional<std::variant<nitro::NullType, std::string>>& /*exercise*/) {
  logDebug("setExercise(): iOS fallback");
}

void HybridTempoClassifier::update(const std::string& /*phase*/, std::optional<double> /*fps*/) {
  logDebug("update(): iOS fallback");
}

std::string HybridTempoClassifier::getCurrentTempo() {
  logDebug("getCurrentTempo(): iOS fallback");
  return "unknown";
}

double HybridTempoClassifier::getCurrentQuality() {
  logDebug("getCurrentQuality(): iOS fallback");
  return 0.0;
}

void HybridTempoClassifier::reset() {
  logDebug("reset(): iOS fallback");
}

} // namespace margelo::nitro::tempoclassifier

File: android/src/main/java/com/margelo/nitro/tempoclassifier/HybridTempoClassifier.kt
package com.margelo.nitro.tempoclassifier

import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.*
import com.margelo.nitro.core.NullType

@DoNotStrip
class HybridTempoClassifier : HybridTempoClassifierSpec() {
    private val TAG = "NitroTempoCls"
    private val engine = TempoClassifierEngine()

    override fun loadModelFromJson(modelJson: String): Boolean {
        Log.d(TAG, "loadModelFromJson(): begin")
        return engine.loadModel(modelJson)
    }

    override fun loadModelFromAsset(assetName: String): Boolean {
        Log.d(TAG, "loadModelFromAsset(): asset=$assetName")
        val context = NitroModules.applicationContext ?: return false
        return try {
            val json = context.assets.open(assetName).bufferedReader().use { it.readText() }
            engine.loadModel(json)
        } catch (e: Exception) {
            Log.e(TAG, "loadModelFromAsset(): failed: ${e.message}")
            false
        }
    }

    override fun setExercise(exercise: Variant_NullType_String?) {
        engine.setExercise(exercise?.asSecondOrNull())
    }

    override fun update(phase: String, fps: Double?) {
        engine.update(phase, fps ?: 0.0)
    }

    override fun getCurrentTempo(): String {
        return engine.getCurrentTempo()
    }

    override fun getCurrentQuality(): Double {
        return engine.getCurrentQuality()
    }

    override fun reset() {
        engine.reset()
    }
}
